packages/react/src/components/dateInput/DateInput.stories.tsx

```diff
@@ -190,5 +191,80 @@ export const WithSelectedDisabledDates = (args) => {
     />
   );
 };
+
 WithSelectedDisabledDates.storyName = 'With selected disabled dates';
 WithSelectedDisabledDates.parameters = { loki: { skip: true } };
+export const WithCustomDayStyles = (args: DateInputProps) => {
+  const dateFormat = 'dd.M.yyyy';
+  const dateValue = new Date(2021, 10, 12);
+  const [value, setValue] = useState<string>(format(dateValue, dateFormat));
+  const helperText = `Custom styles for days with limited available timeslots in the date picker calendar. Use format D.M.YYYY.`;
+
+  const someSpaceLeftDate: LegendItem = {
+    elementId: 'some-space-left',
+    label: 'Some free timeslots available',
+    color: 'var(--color-tram-medium-light)',
+  };
+
+  const littleSpaceLeftDate: LegendItem = {
+    elementId: 'little-space-left',
+    label: 'Only a few free timeslots available',
+    color: 'var(--color-summer-medium-light)',
+  };
+
+  const dayHasSomeSpace = (day: number) => day === 12 || day === 15 || day === 24 || day === 25;
+  const dayHasLittleSpace = (day: number) => day === 3 || day === 4 || day === 17 || day === 19;
+
+  const getDateLegendId = (date: Date) => {
+    const day = date.getDate();
+    if (dayHasSomeSpace(day)) return someSpaceLeftDate.elementId;
+    if (dayHasLittleSpace(day)) return littleSpaceLeftDate.elementId;
+    return undefined;
+  };
+
+  const setDateTheme: (date: Date) => DateCustomTheme | undefined = (date: Date) => {
+    const legendId = getDateLegendId(date);
+    if (legendId === someSpaceLeftDate.elementId)
+      return {
+        '--date-background': someSpaceLeftDate.color,
+      };
+
+    if (legendId === littleSpaceLeftDate.elementId)
+      return {
+        '--date-background': littleSpaceLeftDate.color,
+      };
+    return undefined;
+  };
+
```

Comment:
I think the theme version is better than classNames. Very good change! 👍

Reply:
Too bad each date has own `custom-theme-xxx`, even if colors match. But that's a `useTheme`-problem. And new  styles are created on each time picker is opened. `useTheme` should support fixed styles somehow.

Reply:
I think that theme is better too even with all the troubles it brings. With className and the way DatePicker is built, it proved difficult to override the background-color in all cases. When setting the custom class to Day component, it didn't override, but in the inside span it did. But user may not want to only target the span element. With theme, the options are limited but they work as expected.

Reply:
I limited the DateCustomTheme variables to only `'--date-background'` and `'--date-color'` because only those could override Day component's styles. Seems like the other CSS variables inside `DatePicker.module.scss` should come through a generic `theme` prop, I reckon 🤔

Reply:
Well I changed the implementation back to className because `useTheme` was causing the issues mentioned here and changing months caused an error `Rendered more/fewer hooks than on previous render` because the amount of days per month changed. `setDateClassName` works now with DateInput CSS variables. I tried to style Accordion component's theme variables with className which sets the variables but they aren't available there. I think the reason they work in DateInput is that the `DatePicker.module.scss` declares variables in `:root` which puts them in the root of the component it seems. The docs say it puts the included styles in the root of the DOM, which would be `<html>` but I guess with React they are encapsulated in the component.

Reply:
Customizing dates with css vars via classNames works very well. 👍  

I was curious, if there is an actual need to use :`root` as the base of the css vars for the date input. This is not related to this PR, but if vars are moved from `:root` to `.hds-datepicker` everything still works.

Some of the vars in the `:root` are too generic (also not related to this PR) like `horizontal-spacing,  vertical-spacing,  table-width` and could easily cause conflicts when same names are used somewhere else.

Bottom line: css vars specific to an element should be enclosed to that element, not root. Should we make a new ticket for moving the vars?

Reply:
> Some of the vars in the :root are too generic (also not related to this PR) like horizontal-spacing,  vertical-spacing,  table-width and could easily cause conflicts when same names are used somewhere else.
Bottom line: css vars specific to an element should be enclosed to that element, not root. Should we make a new ticket for moving the vars?

No they are enclosed to that element, not root. You can try to access DatePicker's variable in the story in another element and you'll see that they are not available.

Reply:
> I was curious, if there is an actual need to use :root as the base of the css vars for the date input. This is not related to this PR, but if vars are moved from :root to .hds-datepicker everything still works.

Okay well that's good! I'll add this to the Jira doc I made about theming.

Reply:
> No they are enclosed to that element, not root. You can try to access DatePicker's variable in the story in another element and you'll see that they are not available.

Are they enclosed? I've not used `:root`, but adding a style tag at the end of DOM with new value for `vertical-spacing` alters the DatePicker too.

![image](https://github.com/City-of-Helsinki/helsinki-design-system/assets/2639230/8eadce2e-8581-4497-8a5f-91706f0e1227)

Chrome shows the DatePicker's `:root` values in every element in the inspector. I can use `var(--selected-date-background)` outside DatePicker.

![image](https://github.com/City-of-Helsinki/helsinki-design-system/assets/2639230/5f4281b1-740a-49b6-be1e-be34599244e0)

My previous argument was based on the specs: "The :root [CSS](https://developer.mozilla.org/en-US/docs/Web/CSS) [pseudo-class](https://developer.mozilla.org/en-US/docs/Web/CSS/Pseudo-classes) matches the root element of a tree representing the document."

And I was disappointed that they weren't enclosed. 😄  There is `:scope`, but it is not supported very well.

Reply:
Well, crap. I tested if the variable was available just in a random class, which it wasn't. So the variables are only inside the `:root` selector. That isn't handy of course in case of accidental overriding of existing variables. Well, maybe we don't make moves in this PR but I'll edit the confluence docs.